### PR TITLE
Changed writing style to not use `StopIteration`

### PR DESCRIPTION
**Pull request by haru-44**
I think it would be easier to understand using `for` statements instead of using `StopIteration` in `sqrt_mod`.

Also, I added a docstring for `_sqrt_mod_tonelli_shanks`.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/ntheory/residue_ntheory.py [hunk 1]
```diff
@@ -433,10 +433,34 @@ def _sqrt_mod_tonelli_shanks(a, p):
...
+    Assume that the root exists (if it does not, we are in an infinite loop).
+    Although ``p`` correctly returns the answer for any odd prime,
+    ``p != 1 (mod 8)``, there is no advantage to using this algorithm since
+    a more efficient algorithm exists.
+
+    Parameters
+    ==========
+
+    a : integer
+    p : odd prime number
```
Comment (sylee957) [inline sympy/ntheory/residue_ntheory.py:445]:
```suggestion
    a : int
    p : int
        Odd prime number
```

I now believe that with the sphinx able to pull out the type annotations, it's better to make the type annotation actual python type annotations, and move semantics to details

**Comment by haru-44** [comment]
I don't know the cause, but the test now times out.

File: sympy/ntheory/residue_ntheory.py [hunk 2]
```diff
@@ -493,24 +518,16 @@ def sqrt_mod(a, p, all_roots=False):
...
     if all_roots:
         return sorted(sqrt_mod_iter(a, p))
-    try:
-        p = abs(as_int(p))
-        it = sqrt_mod_iter(a, p)
-        r = next(it)
-        if r > p // 2:
-            return p - r
-        elif r < p // 2:
+    p = abs(as_int(p))
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:521]:
```suggestion
    p = abs(as_int(p))
    halfp = p // 2
```

File: sympy/ntheory/residue_ntheory.py [hunk 3]
```diff
@@ -493,24 +518,16 @@ def sqrt_mod(a, p, all_roots=False):
...
+    x = None
+    for r in sqrt_mod_iter(a, p):
+        if r < p // 2:
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:524]:
```suggestion
        if r < halfp:
```

File: sympy/ntheory/residue_ntheory.py [hunk 4]
```diff
@@ -493,24 +518,16 @@ def sqrt_mod(a, p, all_roots=False):
...
             return r
+        elif r > p // 2:
```
Comment (smichr) [inline sympy/ntheory/residue_ntheory.py:526]:
```suggestion
        elif r > halfp:
```

**Comment by sylee957** [comment]
The test failings are sometimes random

**Comment by haru-44** [comment]
I see. So that's how it happens.

**Comment by haru-44** [comment]
Thanks.

**Pull request merged**